### PR TITLE
Add a number of new codecs plus fixes

### PR DESCRIPTION
**Pull request by pnwpedro**
#### Codecs Included
* BoolCodec
* ByteArrayCodec
* ByteCodec
* CharCodec
* DoubleCodec
* FloatCodec
* LongCodec
* OptionalCodec
* ShortCodec

#### Other Changes/Fixes
* Moves default codec registration into DefaultCodecRegistry singleton
* Create a BYTES token to properly handle @bytes tag (breaking change)
* Handle NULL token in ListCodec, MapCodec, PageCodec
* Handle NULL token in MapCodec
* Handle BYTES token in StringCodec
* Implement equality on a number of types

### Motivation and context

Continue building toward the end state for codecs in this lib!

### How was the change tested?

* Add parameterized codec test to test a number of roundtrip and decode cases
* Add parameterized tagged value test so we won't miss testing for an escapable key

### Change types

* - [X] Bug fix (non-breaking change that fixes an issue)
* - [X] New feature (non-breaking change that adds functionality)
* - [x] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [X] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/fauna/types/NamedDocumentRef.java [hunk 1]
```diff
@@ -55,4 +57,23 @@ public Module getCollection() {
...
+        NamedDocumentRef c = (NamedDocumentRef) o;
+
+        return name.equals(c.name)
+                && getCollection().equals(c.getCollection());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, getCollection());
+    }
```
Comment (freels) [inline src/main/java/com/fauna/types/NamedDocumentRef.java:78]: oh, the java. 😭

**Review by freels — DISMISSED** [review]

File: src/test/java/com/fauna/beans/ClassWithParameterizedFields.java [hunk 2]
```diff
@@ -17,4 +18,24 @@ public class ClassWithParameterizedFields {
 
     @FaunaField(name = "a_map", typeArgument = Integer.class)
     public Map<String,Integer> map  = Map.of("key1", 42);
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+
```
Comment (findgriffin) [inline src/test/java/com/fauna/beans/ClassWithParameterizedFields.java:25]: Is it required that users override the `equals` method like this for all of their classes?

File: src/main/java/com/fauna/codec/codecs/MapCodec.java [hunk 3]
```diff
@@ -21,26 +22,29 @@ public MapCodec(Codec<V> valueCodec) {
...
     @Override
     public L decode(UTF8FaunaParser parser) throws IOException {
-        if (parser.getCurrentTokenType() != FaunaTokenType.START_OBJECT) {
-            throw new ClientException(unexpectedTokenExceptionMessage(parser.getCurrentTokenType()));
-        }
+        switch (parser.getCurrentTokenType()) {
+            case NULL:
+                return null;
+            case START_OBJECT:
+                Map<String, V> map = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/codecs/MapCodec.java:29]: Yeah, I think we should–same with List or enumerables in general–but the complexity is a lot higher. My focus here was do something reasonable for now.

File: src/test/java/com/fauna/beans/ClassWithParameterizedFields.java (hunk 2 above)
Comment (pnwpedro) [reply src/test/java/com/fauna/beans/ClassWithParameterizedFields.java:25]: Nope. This just makes test comparisons easier.

**Review by findgriffin — APPROVED** [review]

**Pull request merged**